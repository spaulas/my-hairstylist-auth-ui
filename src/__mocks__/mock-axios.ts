export const mockPost = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockGet = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockPut = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockDelete = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

export const mockPatch = jest.fn().mockReturnValue({
  data: {},
  callback: jest.fn(),
});

const mockAxios = jest.fn().mockReturnValue({
  post: mockPost,
  get: mockGet,
  put: mockPut,
  patch: mockPatch,
  delete: mockDelete,
});

jest.mock("../config/HttpClient/index.ts", () => ({
  __esModule: true,
  default: mockAxios,
}));
