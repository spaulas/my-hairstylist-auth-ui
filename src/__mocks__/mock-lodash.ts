export const mockDebounce = jest.fn((fn) => console.log('MOCK DEBOUTNCE'));

jest.mock("lodash.debounce", () => ({
  ...jest.requireActual("lodash.debounce"),
  debounce: mockDebounce,
}));
