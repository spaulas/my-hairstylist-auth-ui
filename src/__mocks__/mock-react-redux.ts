import { ExplicitAny } from "@type/global";

export const mockSelector = jest.fn();

export const mockDispatch = jest.fn().mockImplementation((action: ExplicitAny) => null);

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: mockSelector,
    useDispatch: () => mockDispatch
}));
