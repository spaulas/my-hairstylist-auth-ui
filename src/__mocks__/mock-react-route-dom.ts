export const mockPush = jest.fn();
export const mockBlock = jest.fn();
export const mockLocation = jest.fn();

export const mockParams = jest.fn().mockImplementation(() => {});

jest.mock('react-router-dom', () => ({
    useLocation: mockLocation,
    useParams: mockParams,
    useHistory: () => ({
        push: mockPush,
        block: () => ({
            current: mockBlock
        })
    })
}));
