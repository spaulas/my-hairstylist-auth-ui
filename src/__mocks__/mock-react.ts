export const mockUseState = jest.fn();
export const mockUseCallBack = jest.fn((fn) => fn);

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: mockUseState,
  useCallback: mockUseCallBack,
}));
