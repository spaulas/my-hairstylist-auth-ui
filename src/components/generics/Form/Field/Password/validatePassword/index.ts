const validatePassword = (password: string): string => {
  let errorCause = "";

  if (password.length < 8 || password.length > 128) {
    errorCause = "length";
  } else {
    if (!hasAtLeastUpperCaseAndLowerCase(password)) {
      errorCause = "case-";
    }

    if (!hasAtLeastOneNumber(password)) {
      errorCause = `${errorCause}number-`;
    }

    if (!hasAtLeastOneSpecialCharacter(password)) {
      errorCause = `${errorCause}special-`;
    }

    errorCause = errorCause.substring(0, errorCause.length - 1);
  }

  return errorCause;
};

const hasAtLeastUpperCaseAndLowerCase = (password: string) => {
  return /(.*[a-z,A-Z])/.test(password);
};

const hasAtLeastOneNumber = (password: string) => {
  return /(?=.*\d)/.test(password);
};

const hasAtLeastOneSpecialCharacter = (password: string) => {
  return /(?=.*[@$!%*?&#:/<>^=+])/.test(password);
};

export default validatePassword;
