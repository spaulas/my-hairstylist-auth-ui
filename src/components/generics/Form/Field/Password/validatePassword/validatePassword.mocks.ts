export const mockValidatePassword = jest.fn();

jest.mock("./index.ts", () => mockValidatePassword);
