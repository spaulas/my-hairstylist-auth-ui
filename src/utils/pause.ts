export default function pause(seconds: number) {
  return new Promise((r) => setTimeout(r, seconds * 1000));
}
