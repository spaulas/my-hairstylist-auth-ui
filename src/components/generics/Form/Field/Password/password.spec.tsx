import "@testing-library/jest-dom/extend-expect";
import { mockUseState } from "../../../../../__mocks__/mock-react";
import { mockValidatePassword } from "./validatePassword/validatePassword.mocks";
import PasswordField from ".";
import { act, fireEvent, render, screen } from "@testing-library/react";
import pause from "../../../../../utils/pause";

describe("PasswordField", () => {
  const props = {
    password: "",
    setPassword: jest.fn(),
    isFocused: false,
    setIsFocused: jest.fn(),
  };

  const error = "";
  const setErrorMock = jest.fn();

  const changePasswordInput = async (
    newPassword: string,
    errorMessage: string
  ) => {
    mockValidatePassword.mockReturnValue(errorMessage);

    const passwordProps = {
      ...props,
      password: newPassword,
    };
    const { rerender } = render(<PasswordField {...props} />);

    const input = screen.getByTestId("password");
    fireEvent.change(input, { target: { value: newPassword } });

    expect(props.setPassword).toHaveBeenCalledWith(newPassword);

    await act(async () => {
      await pause(1);
    });

    expect(mockValidatePassword).toHaveBeenCalledWith(newPassword);
    expect(setErrorMock).toHaveBeenCalledWith(errorMessage);

    mockUseState.mockReturnValue([errorMessage, setErrorMock]);

    rerender(<PasswordField {...passwordProps} />);

    expect(screen.getByText("Password")).toBeVisible();
    expect(input).toHaveValue(newPassword);
  };

  beforeEach(() => {
    mockUseState.mockReturnValue([error, setErrorMock]);
  });

  it("renders correctly", () => {
    const { container } = render(<PasswordField {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("renders label when there is no password", () => {
    render(<PasswordField {...props} />);
    expect(screen.getByText("Password")).toBeVisible();
  });

  it("renders label and password when there is an password", async () => {
    const password = "P@ssw0rd!";
    await changePasswordInput(password, "");
  });

  describe("when password is invalid shows", () => {
    it("length error message when password is too short", async () => {
      const password = "pa";
      await changePasswordInput(password, "length");

      const errorMessage = screen.getByText(
        "Password should have 8 to 128 characters"
      );
      expect(errorMessage).toBeVisible();
    });

    it("case error message when password has only one case", async () => {
      const password = "password123";
      await changePasswordInput(password, "case");

      const errorMessage = screen.getByText(
        "Password should have at least one uppercase and one lowercase letter"
      );
      expect(errorMessage).toBeVisible();
    });

    it("number error message when password has no numbers", async () => {
      const password = "Password";
      await changePasswordInput(password, "number");

      const errorMessage = screen.getByText(
        "Password should have at least a number"
      );
      expect(errorMessage).toBeVisible();
    });

    it("special error message when password has no special characters", async () => {
      const password = "Password123";
      await changePasswordInput(password, "special");

      const errorMessage = screen.getByText(
        "Password should have at least one special character"
      );
      expect(errorMessage).toBeVisible();
    });

    it("case-number error message when password has no numbers and only one case", async () => {
      const password = "password!";
      await changePasswordInput(password, "case-number");

      const errorMessage = screen.getByText(
        "Password should have at least one uppercase and one lowercase letter, and one number"
      );
      expect(errorMessage).toBeVisible();
    });

    it("case-special error message when password has no special characters and only one case", async () => {
      const password = "password123";
      await changePasswordInput(password, "case-special");

      const errorMessage = screen.getByText(
        "Password should have at least one uppercase and one lowercase letter, and one special character"
      );
      expect(errorMessage).toBeVisible();
    });

    it("number-special error message when password has no special characters and no numbers", async () => {
      const password = "Password";
      await changePasswordInput(password, "number-special");

      const errorMessage = screen.getByText(
        "Password should have at least one number, and one special character"
      );
      expect(errorMessage).toBeVisible();
    });

    it("case-number-special error message when password has no special characters and no numbers", async () => {
      const password = "password";
      await changePasswordInput(password, "case-number-special");

      const errorMessage = screen.getByText(
        "Password should have at least one uppercase and one lowercase letter, one number, and one special character"
      );
      expect(errorMessage).toBeVisible();
    });
  });
});
