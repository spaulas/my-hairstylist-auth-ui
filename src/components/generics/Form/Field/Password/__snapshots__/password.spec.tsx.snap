// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordField renders correctly 1`] = `
<div>
  <div
    class="form--item"
  >
    <p
      class="form--label "
    >
      Password
    </p>
    <input
      class="form--input "
      data-testid="password"
      id="password"
      name="password"
      type="password"
      value=""
    />
    <p
      class="error "
    />
  </div>
</div>
`;
